controllers/vmset_controller.go

```diff
@@ -289,14 +230,129 @@ func (r *VMSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		instance.Status.BaseImageDVReady = true
 	}
 
-	// Create controller VM objects
+	// How many VMs do we currently have for this VMSet?
+	virtualMachineList, err := common.GetAllVirtualMachinesWithLabel(r, map[string]string{
+		OwnerNameLabelSelector: instance.Name,
+	}, instance.Namespace)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	existingVirtualMachines := map[string]string{}
+	removalAnnotatedVirtualMachines := []string{}
+
+	// Generate a map of existing VMs and also store those annotated for potential removal
+	for _, virtualMachine := range virtualMachineList.Items {
+		existingVirtualMachines[virtualMachine.ObjectMeta.Name] = virtualMachine.ObjectMeta.Name
+
+		if val, ok := virtualMachine.Annotations[vmset.VirtualMachineRemovalAnnotation]; ok && (strings.ToLower(val) == "yes" || strings.ToLower(val) == "true") {
+			removalAnnotatedVirtualMachines = append(removalAnnotatedVirtualMachines, virtualMachine.ObjectMeta.Name)
+		}
+	}
+
+	// How many VirtualMachine de-allocations do we need (if any)?
+	oldVmsToRemoveCount := len(existingVirtualMachines) - instance.Spec.VMCount
+
+	if oldVmsToRemoveCount > 0 {
+		oldVmsRemovedCount := 0
+
+		for i := 0; i < oldVmsToRemoveCount; i++ {
+			// First choose VirtualMachines to remove from the prepared list of VirtualMachines
+			// that have the "osp-director.openstack.org/vmset-delete-virtualmachine=yes" annotation
+
+			if len(removalAnnotatedVirtualMachines) > 0 {
```

Comment:
we could also trigger the remove of virtual machines when the removalAnnotatedVirtualMachines number matches the new instance.Spec.VMCount
~~~
if len(removalAnnotatedVirtualMachines) > 0 && len(removalAnnotatedVirtualMachines) ==  instance.Spec.VMCount {
  loop through the removalAnnotatedVirtualMachines
} else {
  otherwise log the warning
}
~~~

with this we only start removing when all vms got annotated to match the new vmcount